### DISCUSSION TITLE
UI freezes with computeWithDatabase

### DISCUSSION THREAD

**Discussion by AnisovAleksey** [simolus3/drift#2455] [Q&A] [answered]
Hello, I use computeWithDatabase to achieve smooth UI experience but noticed little freezes when inserting a lot of data. I tried to analyze performance via CPU Profile and got chart with internal methods of drift database. Could you, please, explain how it works? what actually use CPU and how I can avoid this?
<img width="540" alt="image" src="https://github.com/simolus3/drift/assets/10289829/4d68265b-8d5d-44f1-ba16-37e0642e5bc0">

**Comment by simolus3** [comment]
Could you confirm that what you're seeing is the main isolate? This is all drift, `Sqlite3Delegate.runBatchSync` is responsible for running the statements (which happens synchronously and is generally CPU/IO intensive).

How are you constructing the database, are you using `NativeDatabase.createInBackground` or a regular `NativeDatabase`?

**Reply by AnisovAleksey** [reply]
regular `NativeDatabase`, and also with encryption.

```
return NativeDatabase(
      file,
      setup: (db) {
        assert(() {
          return db.select('PRAGMA cipher_version;').isNotEmpty;
        }());

        db.execute("pragma key = '${String.fromCharCodes(databaseKey)}'");

        db.execute("PRAGMA foreign_keys = ON;");
      },
    );
```

yes, this from main isolate. I've exported profile data, you can look it in dev tools 
[dart_devtools_2023-06-03_11_04_15.401.txt](https://github.com/simolus3/drift/files/11641696/dart_devtools_2023-06-03_11_04_15.401.txt) but rename it to **.json**

**Reply by AnisovAleksey** [reply]
NativeDatabase.createInBackground doesn't work with encrypted database, I get the following error
```
E/flutter ( 3712): [ERROR:flutter/runtime/dart_vm_initializer.cc(41)] Unhandled Exception: Invalid argument(s): Failed to load dynamic library '/data/data/com.example/lib/libsqlite3.so': dlopen failed: library "/data/data/com.example/lib/libsqlite3.so" not found
E/flutter ( 3712): #0      DriftCommunication.request (package:drift/src/remote/communication.dart:113)
E/flutter ( 3712): #1      _RemoteQueryExecutor.ensureOpen (package:drift/src/remote/client_impl.dart:158)
E/flutter ( 3712): #2      LazyDatabase.ensureOpen.<anonymous closure> (package:drift/src/utils/lazy_database.dart:61)
E/flutter ( 3712): <asynchronous suspension>
E/flutter ( 3712): 
```

**Reply by simolus3** [reply, marked as answer]
> regular `NativeDatabase`, and also with encryption.

The documentation for `computeWithDatabase` does a poor job of explaining this, but that method doesn't create a second database connection. It just "shares" the existing connection with a freshly spawned isolate. This is a problem in your case because you're using a synchronous `NativeDatabase`.

So the expensive batches you're constructing on a short-lived isolate will actually get sent back to the main isolate to be executed synchronously on there. To benefit from this, I recommend using a longer-lived isolate for the entire database (I'll share how to do that with encryption as well).

You might not need `computeWithDatabase` at all when using an encrypted database hosted by a background isolate. There are essentially two things that are (potentially) expensive in drift:

1. Running SQL statements (mostly due to IO, but also potentially CPU-heavy depending on the query).
2. Constructing SQL statements with drift's Dart API and mapping results (can be CPU-heavy).

A background database (like `NativeDatabase.createInBackground`) solves the first issue by transparently running SQL on a background isolate.
As drift's query builder API is synchronous, `computeWithDatabase` is there for the rare cases where step 2. is expensive as well. It also works well for cases where a batch is loaded from network with `Isolate.run`.

> NativeDatabase.createInBackground doesn't work with encrypted database, I get the following error

Ah right, the issue is that the open override needs to be installed on the isolate creating the native database. You can use something like this to create your encrypted `NativeDatabase` in a background isolate:

```dart
QueryExecutor createEncryptedDatabaseInBackground() {
  return DatabaseConnection.delayed(Future(() async {
    final file = /* pick database file */;

    final isolate = await DriftIsolate.spawn(() {
      setupSqlCipher(); // Make sure we're using sqlcipher.so
      return NativeDatabase(
        file,
        setup: (db) {
          assert(() {
            return db.select('PRAGMA cipher_version;').isNotEmpty;
          }());

          db.execute("pragma key = '${String.fromCharCodes(databaseKey)}'");

          db.execute("PRAGMA foreign_keys = ON;");
        },
      );
    });

    return await isolate.connect(
      singleClientMode: true,
    );
  }));
}
```

**Reply by AnisovAleksey** [reply]
wow, thanks for that code, now it worked and freezes are gone 🎉
